### ISSUE TITLE
closure stackoverflow error with :js-provider :shadow

### ISSUE THREAD

**Issue by philippkueng** [thheller/shadow-cljs#509]
Hello everyone,

I'm working on a pet project attempting to have a full-text search engine in the browser and as part of that need 2 different npm modules integrated. Using both dependencies leads to a stackoverflow when using `:js-options {:js-provider :shadow}}`, when using `:js-options {:js-provider :require}}` it works but it's unusable for when targeting the browser. Also using only one of those dependencies at a time also works and funnily enough every now and then the compilation succeeds without running into a stackoverflow error (using a new macbook pro, so resource constraints shouldn't really be an issue). The error I'm getting is below:

```
> shadow-cljs compile app

shadow-cljs - updating dependencies
shadow-cljs - dependencies updated
[:app] Compiling ...

> minimal-shadow-cljs-browser@1.0.0 compile /Users/philippkueng/Documents/bitfondue/closure-stackoverflow-error
> shadow-cljs compile app

shadow-cljs - config: /Users/philippkueng/Documents/bitfondue/closure-stackoverflow-error/shadow-cljs.edn  cli version: 2.8.39  node: v12.1.0
shadow-cljs - updating dependencies
shadow-cljs - dependencies updated
[:app] Compiling ...
Exception in thread "main" java.lang.StackOverflowError
	at com.google.javascript.jscomp.RemoveUnusedCode.traverseChildren(RemoveUnusedCode.java:1186)
	at com.google.javascript.jscomp.RemoveUnusedCode.traverseNode(RemoveUnusedCode.java:436)
	at com.google.javascript.jscomp.RemoveUnusedCode.traverseChildren(RemoveUnusedCode.java:1186)
	at com.google.javascript.jscomp.RemoveUnusedCode.traverseNode(RemoveUnusedCode.java:436)
	at com.google.javascript.jscomp.RemoveUnusedCode.traverseChildren(RemoveUnusedCode.java:1186)
	at com.google.javascript.jscomp.RemoveUnusedCode.traverseNode(RemoveUnusedCode.java:436)
	at com.google.javascript.jscomp.RemoveUnusedCode.traverseChildren(RemoveUnusedCode.java:1186)
	at com.google.javascript.jscomp.RemoveUnusedCode.traverseNode(RemoveUnusedCode.java:436)
	at com.google.javascript.jscomp.RemoveUnusedCode.traverseChildren(RemoveUnusedCode.java:1186)
	at com.google.javascript.jscomp.RemoveUnusedCode.traverseNode(RemoveUnusedCode.java:436)
```

Currrently looking through the shadow-cljs source code attempting to just optimize the npm modules for whitespace and see if that fixes it rather than using a minified version. Also looking in the closure repository I couldn't find anything similar yet.

If any of you reading this has some pointers of things I could attempt, please let me know.

I created a repo which contains a reproduction of the error here https://github.com/philippkueng/closure-stackoverflow-error

**Comment by thheller** [comment]
That is odd. The bug seems to appear in the `lucene-tokenizers` package. Should be possible to reproduce this with just the Closure Compiler since it doesn't have any other dependencies.

Whitespace-only is not an option since we need the dead-code removal simple does to get rid of conditional requires in other packages.

**Comment by philippkueng** [comment]
You were right as that `lucene-tokenizers` contained the offending code.

First I cloned the `master` branch of https://github.com/google/closure-compiler and checked out the version `v20190325` which I believe is closest to the version being used within shadow-cljs.

Then ran the offending javascript file through the closure-compiler 

```
java -jar target/closure-compiler-1.0-SNAPSHOT.jar --js_output_file=out.js /Users/philippkueng/closure-stackoverflow-error/node_modules/lucene-tokenizers/lucene-tokenizers.babel.js
```

which resulted in no output whatsoever

I then copied this new `out.js` to the `tokenizers.babel.js` location `cp out.js /Users/philippkueng/closure-stackoverflow-error/node_modules/lucene-tokenizers/lucene-tokenizers.babel.js` and re-ran `npm run compile`. Which succeeded this time without any warnings or errors.

Furthermore I tried doing the same with dexie, leaving `lucene-tokenizers` as it is coming from npmjs.org, this however resulted in the same error reported in the issue above (stackoverflow error).

My guess from this is that there's something broken with the one javascript file in `lucene-tokenizers` which the closure-compiler can fix when just working with a single file but is unable to address when working with multiple files.

@thheller I'll close this issue for now as my immediate problem has a workaround but will hopefully address it in the coming weeks and update the issue with new findings. Thank you very much for putting me on the right path and shadow-cljs to begin with! Truly a joy to work with.
